### PR TITLE
NEW Allow developers to exclude LinkText field

### PR DESCRIPTION
**Pull request by GuySartorelli**
Because this gets passed from PHP through to JS, back to PHP for the controller, and then finally through to JS _again_ for the form modal, I think the only way to adequately test this is through behat tests. There's a card in progress right now to create behat tests, so I'm not going to duplicate efforts creating a behat setup in this PR.

If the reviewer agrees, I'll create a separate card to add behat tests for this scenario, which can be blocked on the existing behat card.

## Issue
- https://github.com/silverstripe/silverstripe-linkfield/issues/211

### CODE REVIEW

File: src/Form/Traits/AllowedLinkClassesTrait.php [hunk 1]
```diff
@@ -15,6 +15,19 @@ trait AllowedLinkClassesTrait
...
+    public function setExcludeLinkTextField(bool $include): static
+    {
+        $this->excludeLinkTextField = $include;
+        return $this;
+    }
+
+    public function getExcludeLinkTextField(): bool
+    {
+        return $this->excludeLinkTextField;
+    }
```
Comment (ChloeSartorelli) [inline src/Form/Traits/AllowedLinkClassesTrait.php:29]: I didn't want to create a whole separate trait just for this, nor did I want to duplicate code by putting it into each field.

I've added a note to https://github.com/silverstripe/silverstripe-linkfield/issues/78 to consider combining the traits into an abstract class instead - assuming that is done, it doesn't matter _too_ much where this code goes for this PR since it'll all end up in the same place anyway.

File: src/Controllers/LinkFieldController.php [hunk 2]
```diff
@@ -88,7 +88,7 @@ public function linkForm(): Form
             }
             $operation = 'create';
         }
-        return $this->createLinkForm($link, $operation);
+        return $this->createLinkForm($link, $operation, (bool) $this->getRequest()->getVar('excludeTextField'));
```
Comment (emteknetnz) [inline src/Controllers/LinkFieldController.php]: Bit messy having this one on line, create an intermediate variable.

File: client/src/components/LinkField/LinkField.js [hunk 3]
```diff
@@ -52,6 +52,7 @@ const LinkField = ({
   ownerID,
   ownerClass,
   ownerRelation,
+  excludeTextField = false,
```
Comment (emteknetnz) [inline client/src/components/LinkField/LinkField.js]:
```suggestion
  excludeLinkTextField = false,
```
It's really confusing omitting the `Link` from this variable, especially since in other places it's not omitted.  It's a specific field we want to remove, not just a generic 'textfield'

Update all variables through including the ones passed to jquery as an element attribute e.g. 'data-exclude-textfield' should be 'data-exclude-linktext-field

**Review by emteknetnz — CHANGES_REQUESTED** [review]
Add a unit test on LinkFieldController that passing an 'excludeLinkTextField' get param will result in a schema without a LinkText field

File: src/Controllers/LinkFieldController.php (hunk 2 above)
Comment (ChloeSartorelli) [reply src/Controllers/LinkFieldController.php]: I'd argue having the intermediate variable is unnecessary 😅 Gotta love subjective code styles.
I've made the change.

File: client/src/components/LinkField/LinkField.js (hunk 3 above)
Comment (ChloeSartorelli) [reply client/src/components/LinkField/LinkField.js]: Done

**Comment by GuySartorelli** [comment]
> Add a unit test on LinkFieldController that passing an 'excludeLinkTextField' get param will result in a schema without a LinkText field

Done

File: tests/php/Controllers/LinkFieldControllerTest.php [hunk 4]
```diff
@@ -2,12 +2,13 @@
 
 namespace SilverStripe\LinkField\Tests\Controllers;
 
-use SilverStripe\CMS\Model\SiteTree;
+use ReflectionMethod;
 use SilverStripe\Dev\FunctionalTest;
 use SilverStripe\LinkField\Tests\Controllers\LinkFieldControllerTest\TestPhoneLink;
-use SilverStripe\Core\Config\Config;
```
Comment (ChloeSartorelli) [inline tests/php/Controllers/LinkFieldControllerTest.php:8]: `use` statements that were removed are unrelated refactoring - those were not in use.

File: client/src/entwine/LinkField.js [hunk 5]
```diff
@@ -51,6 +51,7 @@ jQuery.entwine('ss', ($) => {
         ownerID: inputField.data('owner-id'),
         ownerClass: inputField.data('owner-class'),
         ownerRelation: inputField.data('owner-relation'),
+        excludeLinkTextField: inputField.data('exclude-textfield'),
```
Comment (emteknetnz) [inline client/src/entwine/LinkField.js]:
```suggestion
        excludeLinkTextField: inputField.data('exclude-linktext-field'),
```

File: src/Form/LinkField.php [hunk 6]
```diff
@@ -51,6 +51,7 @@ protected function getDefaultAttributes(): array
         $attributes['data-owner-id'] = $ownerFields['ID'];
         $attributes['data-owner-class'] = $ownerFields['Class'];
         $attributes['data-owner-relation'] = $ownerFields['Relation'];
+        $attributes['data-exclude-textfield'] = $this->getExcludeLinkTextField();
```
Comment (emteknetnz) [inline src/Form/LinkField.php]:
```suggestion
        $attributes['data-exclude-linktext-field'] = $this->getExcludeLinkTextField();
```

File: src/Form/MultiLinkField.php [hunk 7]
```diff
@@ -70,6 +71,7 @@ protected function getDefaultAttributes(): array
         $attributes['data-owner-id'] = $ownerFields['ID'];
         $attributes['data-owner-class'] = $ownerFields['Class'];
         $attributes['data-owner-relation'] = $ownerFields['Relation'];
+        $attributes['data-exclude-textfield'] = $this->getExcludeLinkTextField();
```
Comment (emteknetnz) [inline src/Form/MultiLinkField.php]:
```suggestion
        $attributes['data-exclude-linktext-field'] = $this->getExcludeLinkTextField();
```

**Review by emteknetnz — CHANGES_REQUESTED** [review]

File: client/src/entwine/LinkField.js (hunk 5 above)
Comment (ChloeSartorelli) [reply client/src/entwine/LinkField.js]: Oh yeah oops. Fixed.

**Review by emteknetnz — APPROVED** [review]
Tested locally works good
